### PR TITLE
The hybrid learning home page

### LINKED ISSUE #8137: Create first version of Learn > Home tab
CleanShot 2021-06-02 at 13 45 21

CleanShot 2021-06-02 at 14 00 22

Implementation may use current version of ContentCard. For items 2,5,6 , the components created in #8136 should be used 

Items:

1. Create "Home" tab 
2. Add "Your classes" section
3. Link "View all" to AllClassesPage
4. Add "Continue learning on your own" (For now, just show items that are currently in Channels > Resume)
5. Add "Recent lessons section" (For now just show the first 3 lessons, then link "View all" to Class Assignments Page)
6. Add "Recent quizzes section" (Same as above)
7. Add "Explore channels" section (Let "View all" link to current "Channels" tab)
8. Implement the "Guest version" of this page

## Follow up issues

1. Enforce granular device settings, esp. re being able to access un-assigned content
2. White background for corebase

### PR DESCRIPTION
**Pull request by MisRob**
## Summary

Add a new home page. It doesn't follow the full spec in Figma but is rather an MVP based on #8137. See test scenarios below for details. This PR is meant to be reviewed together with #8430, both code and testing, which contains the majority of work and was merged prematurely due to the string freeze. You're welcome to add review comments to already merged #8430.

## References

- closes #8137
- [full feature spec in Figma](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=1788%3A6234)
- [home page cards designs in Figma](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=114%3A8916)

## Reviewer guidance

See [the following test scenarios in Notion](https://learningequality.notion.site/Home-Page-QA-8ffb1011fa034e21bc1f52a2aad585ac)

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [x] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by MisRob** [comment]
I am not sure why the linting check doesn't pass, will investigate later

File: kolibri/plugins/learn/assets/src/composables/useChannels.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/**
+ * A composable function containing logic related to channels
+ */
+
+import { ref } from 'kolibri.lib.vueCompositionApi';
+import { set } from '@vueuse/core';
+import { ChannelResource } from 'kolibri.resources';
+
+// The refs are defined in the outer scope so they can be use as a store
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/composables/useChannels.js]:
```suggestion
// The refs are defined in the outer scope so they can be used as a shared store
```

File: kolibri/plugins/learn/assets/src/views/cards/CardGrid.vue [hunk 2]
```diff
@@ -11,14 +11,31 @@
...
+       *   Level 2: 2 cards
+       *   Level 1: 1 cards
+       *   Level 0: 1 card
+       *
+       * Grid type `2`
+       *   Level 3+: 4 cards
+       *   Level 2: 3 cards
+       *   Level 1: 2 cards
+       *   Level 0: 1 card
+       */
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/cards/CardGrid.vue:38]: thanks for adding this comment. Could also possibly add a validator like

```js
validator(value) {
  return [GRID_TYPE_1, GRID_TYPE_2].includes(value);
}
```

**Review by indirectlylit — COMMENTED** [review]
Intriguing use of composable as a vuex replacement - I hadn't seen this before. Do you like it better than vuex?

**Comment by MisRob** [comment]
@jtamiace's feedback:

> could we use text.annotation  as the color of class names in these cards?
screen_shot_2021-10-19_at_6 17 28_pm


Will update

**Comment by MisRob** [comment]
In "Continue learning on your own", a topic title that should be displayed above a resource title is missing (it's implemented so I might introduce a bug with the most recent updates)

Screenshot from 2021-10-21 15-41-42

**Comment by MisRob** [comment]
@rtibbles As for API requests that the new home page is responsible for triggering, they're defined [in this handler](https://github.com/learningequality/kolibri/pull/8504/files#diff-cdb2ba5c6ec7d704c3556d3c2ea149c954b0331732dcb422d349286de8421e0aL53) and for example, for a signed-in user (who needs more data than a guest) it's four calls to these API endpoints:

- /api/content/channel/?available=true
- /learn/api/learnerclassroom/
- /api/content/contentnode/<user_id>/resume/
- /api/content/contentnodeprogress/

**Comment by pcenov** [comment]
Hi @MisRob, I spent some time yesterday and today testing the [buildkite builds](https://buildkite.com/learningequality/kolibri-python-package/builds/3900) on Win10 and Ubuntu, went through all of the neatly written [Notion scenarios](https://learningequality.notion.site/Home-Page-QA-8ffb1011fa034e21bc1f52a2aad585ac) and I'm happy to report that I didn't uncover any issues with the Home Page or the sections which are displayed depending on various variables such as whether the user is signed in or not and whether the user is enrolled in classes or not. Let me know when you fix the issue with the missing topic title in "Continue learning on your own" and I'll check for regression issues.

**Comment by MisRob** [comment]
Annotation text color is now properly used for a class/topic name on cards
Screenshot from 2021-10-25 13-19-08 (copy)
Screenshot from 2021-10-25 13-19-27
Screenshot from 2021-10-25 13-19-34
Screenshot from 2021-10-25 13-19-39
:

**Comment by MisRob** [comment]
Missing topic title in "Continue learning on your own" has been fixed

Screenshot from 2021-10-25 13-19-08

cc @pcenov

**Comment by MisRob** [comment]
@indirectlylit 

> Intriguing use of composable as a vuex replacement - I hadn't seen this before. Do you like it better than vuex?

Yes, I'd say it has brought lots of clarity into the home page logic and writing tests even for more complex views that need to use composables feels much simpler than for those utilizing the whole Vuex machinery. Feels like a step towards functional programming with reactivity advantages. I am not sure how it compares to things that need to be handled accross plugins and would like to get a bit more experience - this was the first time I used it in production code since originally I was building it on top of JB's work who started using composables for the home page. If there's anything specific you're curious about, happy to chat more.

**Comment by MisRob** [comment]
While testing something else in IE11, I've noticed two new problems:

1. **Text ellipsis characters showing on cards even when a title is not overflowing:**
bug-2

2. **Broken thumbnails on channel cards:** Might be caused by flex being buggy in IE. I used new  channel cards with simpler thumbnails styling but until we get to merging our old and new cards components, it may be better to copy [the original styling](https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/plugins/learn/assets/src/views/ChannelCard.vue#L140) that doesn't seem to break in this way
bug-1

**Comment by MisRob** [comment]
Tech debt issues related to this PR as well:

- #8532
- #8533

**Comment by MisRob** [comment]
> While testing something else in IE11, I've noticed two new problems:
>
>  1. Text ellipsis characters showing on cards even when a title is not overflowing
>  2. Broken thumbnails on channel cards

Fixed

**Comment by MisRob** [comment]
As far as I can remember, these should be two last tech debt or improvement issues related to the home page:

- https://github.com/learningequality/kolibri-design-system/issues/267
- https://github.com/learningequality/kolibri-design-system/issues/268

**Comment by MisRob** [comment]
@pcenov Thank you for testing. I finished all updates so if you'd like to check it out, I'll appreciate it. I haven't changed any logic related to what sections are displayed, it was rather some smaller UI tweaks that you can read more about in the comments above, so I think it'd be helpful to focus on cards in different resolutions across browsers, especially in IE.

**Comment by pcenov** [comment]
Hi @MisRob all looks very well, I tested in several browsers (including IE11) and different resolutions and was not able to identify any new issues.
Just one question though: there are no time/progress indicators in the 'Continue learning on your own' cards as indicated in Figma. Is this covered in a separate PR?

**Review by nucleogenesis — COMMENTED** [review]
The code looks brilliant. Excellent work with thorough tests, JSDoc comments <3, excellent clarifying comments and good separation of concerns. I enjoyed getting a deeper dive into a and non-trivial real-world composition API implementation!

Do we need these breadcrumbs? Are they in scope here or will they be fixed in follow up?

screenshot-localhost_8000-2021 10 27-12_38_53

I'm also seeing things that are off with the card styling where the thing under the ActivityType is centered, which looks odd - see "reference" is unaligned with anything else.

screenshot-localhost_8000-2021 10 27-12_51_47

**Comment by rtibbles** [comment]
> Do we need these breadcrumbs? Are they in scope here or will they be fixed in follow up?

I think the errant breadcrumbs are fixed in https://github.com/learningequality/kolibri/pull/8548, so we can ignore for now.

**Review by rtibbles — APPROVED** [review]
Follow up issues filed, and manual testing checks out.

Code read through by @nucleogenesis and a quick skim from me raised no red flags.

**Pull request merged**